src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -22127,6 +22127,1449 @@ void Main(string? s)
+                //         _ = F1<T?>(t5).Value; // 17
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "F1<T?>(t5)").WithLocation(43, 13),
+                // (44,13): warning CS8629: Nullable value type may be null.
+                //         _ = F5<T>(t5).Value; // 18
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "F5<T>(t5)").WithLocation(44, 13));
+        }
+
+        [Fact]
+        public void MaybeNull_ReturnValue_03()
+        {
+            var source =
+@"using System.Runtime.CompilerServices;
+class Program
+{
+    [return: MaybeNull] static string F1() => string.Empty;
+    [return: MaybeNull] static string? F2() => string.Empty;
+    static void M()
+    {
+        F1().ToString(); // 1
+        F2().ToString(); // 2
+    }
+}";
+            var comp = CreateCompilation(new[] { MaybeNullAttributeDefinition, source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (8,9): warning CS8602: Dereference of a possibly null reference.
+                //         F1().ToString(); // 1
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F1()").WithLocation(8, 9),
+                // (9,9): warning CS8602: Dereference of a possibly null reference.
+                //         F2().ToString(); // 2
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "F2()").WithLocation(9, 9));
+        }
+
+        [Fact]
+        public void MaybeNull_ReturnValue_04()
+        {
+            var source =
+@"using System.Runtime.CompilerServices;
+class Program
+{
+    [return: MaybeNull] static int F1() => 1;
+    [return: MaybeNull] static int? F2() => 2;
+    static void M()
+    {
+        F1().ToString();
+        _ = F2().Value; // 1
+    }
+}";
+            var comp = CreateCompilation(new[] { MaybeNullAttributeDefinition, source }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics(
+                // (9,13): warning CS8629: Nullable value type may be null.
+                //         _ = F2().Value; // 1
+                Diagnostic(ErrorCode.WRN_NullableValueTypeMayBeNull, "F2()").WithLocation(9, 13));
+        }
+
+        [Fact]
+        public void MaybeNull_ReturnValue_05()
+        {
+            var source0 =
+@"using System.Runtime.CompilerServices;
+public class A
+{
+    public static T F1<T>(T t) where T : class => t;
+    [return: MaybeNull] public static T F2<T>(T t) where T : class => t;
+}";
+            var comp = CreateCompilation(new[] { MaybeNullAttributeDefinition, source0 }, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics();
+            var ref0 = comp.EmitToImageReference();
+
+            var source =
+@"class B : A
+{
+    static void Main()
+    {
+        object x = null; // 1
+        object? y = new object();
+        F1(x).ToString(); // 2
+        F1(y).ToString();
+        F2(x).ToString(); // 3
+        F2(y).ToString(); // 4
```

Comment:
> // 4 [](start = 25, length = 5)

:memo: incorrect comment

Reply:
I'm not sure I agree.  The attribute was meaningless where written, as the return type could have been written `T?`.  So I think it is safe to ignore it in analysis.

---
In reply to: [288763015](https://github.com/dotnet/roslyn/pull/35955#discussion_r288763015) [](ancestors = 288763015)

Reply:
Having said that, we should have more tests for reading from assembly images with meaningful annotations.

---
In reply to: [289210231](https://github.com/dotnet/roslyn/pull/35955#discussion_r289210231) [](ancestors = 289210231,288763015)